Fix volume seg async save

Punzo opened the pull request:
See https://github.com/Slicer/Slicer/pull/8141

Punzo commented:
@lassoan

I'm wondering if we should add a common method to the Slicer app logic for safer file writing, particularly to handle concurrent write scenarios more robustly.

Instead of using the approach proposed in this PR—creating a temporary folder and then overwriting files—it might be more reliable to implement a **file-based locking mechanism**.

There are two main use cases for this:

1. **Async node bulk data saving**  
   For example, in [this commit](https://github.com/Slicer/Slicer/pull/8141/commits/890c6376d84aaf07372bf3dd58e8a13f17be6297), async saving can conflict with Slicer's standard synchronous save mechanism.

2. **Multiple concurrent Slicer sessions**  
   When multiple sessions are writing (either async or sync), they may attempt to write to the same file or directory concurrently, leading to race conditions or data corruption.

A centralized, locking-based utility for file output could make these operations more robust and prevent such conflicts.

Would this be a direction worth exploring?

pieper commented:
I would love to see a consistent approach to concurrency in the slicer core libraries, and starting with IO makes a lot of sense.  I had done experiments before with lots of model files (dti tractography vtk polydata) and got it working with temporary scenes per-model node that were then merged.  It worked, but it didn't save a lot of time in general cases.  

I think we should profile common use cases to identify bottlenecks.  I know that for large volumes gzip deflating is a single-thread bottleneck, so if compressing .seg.nrrds is in the critical path then spawning it off makes sense.  On the other hand I've been thinking that the zarr of compressing subblocks in parallel is more scalable and a bit cleaner.

Punzo commented:
@lassoan by combining a lock file-based approach with writing files to a unique temporary folder and then performing an atomic rename,  the issue of file corruption during concurrent writes is resolved. 

This PR is ready for review. 

NOTE: this review address the issue only for the volumes and segmentation storage nodes for the moment.

lassoan reviewed (commented):
This will be good, but we can probably simplify the writing preparation (just create a unique temporary folder) and then we should use just a single lockfile to serialize the moving of written files to their final location.

I'm also wondering why this is only done for segmentation and volume files and not for others. Is it only a problem if multiple files are created (.obj files are written as multiple files, too)? Or if writing takes long time (transforms and sequences can take a long time to write, too)? If they should be implemented for other node types, too, but right now they are just not implemented due to lack of time then this should be documented (maybe enough to describe in the commit comment).

Add more details in the commit message: give a high-level overview of what is done and why.

Punzo commented:
> This will be good, but we can probably simplify the writing preparation (just create a unique temporary folder) and then we should use just a single lockfile to serialize the moving of written files to their final location.
> 
> I'm also wondering why this is only done for segmentation and volume files and not for others. Is it only a problem if multiple files are created (.obj files are written as multiple files, too)? Or if writing takes long time (transforms and sequences can take a long time to write, too)? If they should be implemented for other node types, too, but right now they are just not implemented due to lack of time then this should be documented (maybe enough to describe in the commit comment).
> 
> Add more details in the commit message: give a high-level overview of what is done and why.

Thank you for your review and suggestions. I agree that this approach should also be extended to models, sequences, and non-linear transforms. Due to development time constraints, the current implementation is limited to segmentations and volumes. My plan is to implement similar handling for the other node types in the future. Additionally, starting with a couple of classes allows us to test and validate the approach before generalizing it further.

I will address the in-line comments and update the commit message with a more detailed high-level overview tomorrow.

Punzo commented on Libs/MRML/Core/vtkMRMLStorageNode.cxx:
```diff
@@ -1648,3 +1661,223 @@ void vtkMRMLStorageNode::FixFileName()
 }
+
+//----------------------------------------------------------------------------
+bool vtkMRMLStorageNode::CreateUniqueTempFilesForWrite(vtkStringArray* fullNames, std::string& tempDir, vtkStringArray* tempFiles)
```
the method was actually just generating the temp folder and not the temporary files. I cleaned the method (renamed, changed input/output). See new implementation in rebased commit.

Punzo commented on Libs/MRML/Core/vtkMRMLStorageNode.cxx:
```diff
@@ -1648,3 +1661,223 @@ void vtkMRMLStorageNode::FixFileName()
+  if (sourceDir.empty())
+  {
+    vtkErrorToMessageCollectionMacro(this->GetUserMessages(), "vtkMRMLStorageNode::MoveFilesWithLocking", "Source directory string is empty. Cannot commit files.");
+    vtksys::SystemTools::RemoveADirectory(sourceDir.c_str());
```
ok removed the remove of the temp directory and moved in the calling methods. See new implementation in rebased commit.

Punzo commented on Libs/MRML/Core/vtkMRMLStorageNode.cxx:
```diff
@@ -1648,3 +1661,223 @@ void vtkMRMLStorageNode::FixFileName()
+  tempFiles->Reset();
+  for (vtkIdType fileIndex = 0; fileIndex < fullNames->GetNumberOfValues(); ++fileIndex)
+  {
+    std::string fileName = vtksys::SystemTools::GetFilenameName(fullNames->GetValue(fileIndex));
```
the method was actually just generating the temp folder and not the temporary files. I cleaned the method (renamed, changed input/output). See new implementation in rebased commit.

Punzo commented:
@lassoan I have applied your review in the rebased commit https://github.com/Slicer/Slicer/pull/8450/commits/3c6ad5f41a2cf48bcc334c4d3c0d666bd0477a14

I have also updated to latest Slicer and added tests.
